app/controllers/spree/komoju_controller.rb

```diff
@@ -7,18 +7,55 @@ def callback
 
       case params[:type]
       when "payment.captured"
-        order_number = extract_payment_number(params[:data][:external_order_num])
-        payment = Spree::Payment.find_by_number!(order_number)
-        payment.complete! unless payment.completed?
+        payment_captured!
+      when "payment.refunded"
+        payment_refunded!
       end
 
       head 200
     end
 
     private
 
-    def extract_payment_number(external_order_num)
-      external_order_num.split('-').try(:last)
+    def order_number
+      params[:data][:external_order_num].split("-").try(:first)
+    end
+
+    def payment_number
+      params[:data][:external_order_num].split("-").try(:last)
+    end
+
+    def order
+      @order ||= Spree::Order.find_by_number!(order_number)
+    end
+
+    def payment
+      @payment ||= order.payments.find_by_number!(payment_number)
+    end
+
+    def payment_captured!
+      payment.complete! unless payment.completed?
+    end
+
+    def refund_params
+      params[:data][:refunds].last
+    end
+
+    def refund_description
+      refund_params[:description].blank? ? "Komoju refund" : refund_params[:description]
+    end
+
+    def refund_amount
+      # Converts cents into dollars
+      ::Money.new(refund_params[:amount], "USD").to_f
+    end
+
+    def payment_refunded!
+      if payment.completed? && payment.credit_allowed >= refund_amount && payment.currency == refund_params[:currency]
```

Comment:
Since `payment.completed?` never changes as you said I think we just need `if payment.credit_allowed >= refund_amount` here. I don't know a case where a refund would use the wrong currency, it's exception, we can raise an error in that case.